frontend/src/components/PokemonCardDetectorComponent.tsx

```diff
@@ -0,0 +1,585 @@
+  imageUrl: string
+  confidence: number
+  hash?: string
+  matchedCard?: {
+    id: string
+    distance: number
+    imageUrl?: string
+  }
+  topMatches?: Array<{
+    id: string
+    distance: number
+    card: Card
+  }>
+  selected?: boolean
+}
+
+const PokemonCardDetector: React.FC<PokemonCardDetectorProps> = ({ onDetectionComplete, modelPath = '/model/model.json' }) => {
+  const { user } = use(UserContext)
+  const { ownedCards, setOwnedCards } = use(CollectionContext)
+  const [isOpen, setIsOpen] = useState(false)
+  const [_images, setImages] = useState<File[]>([])
+  const [isLoading, setIsLoading] = useState<boolean>(false)
+  const [_results, setResults] = useState<DetectionResult[]>([])
+  const [extractedCards, setExtractedCards] = useState<ExtractedCard[]>([])
+  const [amount, setAmount] = useState(1)
+  const [isProcessing, setIsProcessing] = useState(false)
+  const [isGeneratingHashes, setIsGeneratingHashes] = useState(false)
+  const [showPotentialMatches, setShowPotentialMatches] = useState(false)
+  const fileInputRef = useRef<HTMLInputElement>(null)
+  const detectorService = PokemonCardDetectorService.getInstance()
+
+  useEffect(() => {
+    const initializeModel = async () => {
+      try {
+        setIsLoading(true)
+        await detectorService.loadModel(modelPath)
+        console.log('Model loaded successfully')
+      } catch (error) {
+        console.error('Error loading model:', error)
+      } finally {
+        setIsLoading(false)
+      }
+    }
+
+    initializeModel()
+  }, [modelPath])
+
+  const hashingService = ImageSimilarityService.getInstance()
+  const hashStorageService = CardHashStorageService.getInstance()
+  const uniqueCards = useMemo(() => {
+    return allCards.reduce((acc, card) => {
+      if (!acc.some((c) => c.card_id === card.card_id)) {
+        acc.push(card)
+      }
+      return acc
+    }, [] as Card[])
+  }, [allCards])
+  useEffect(() => {
+    const generateAndStoreHashes = async () => {
+      try {
+        setIsGeneratingHashes(true)
+        await hashStorageService.initDB()
+        const storedHashCount = await hashStorageService.getHashCount()
+
+        if (storedHashCount !== uniqueCards.length) {
+          console.log('Generating hashes for all cards...')
+
+          const batchSize = 80
+          const totalCards = uniqueCards.length
+          const allHashes = []
+
+          for (let i = 0; i < totalCards; i += batchSize) {
+            const batch = uniqueCards.slice(i, i + batchSize)
+
+            // Process one batch
+            const batchPromises = batch.map(async (card) => {
+              try {
+                const hash = await hashingService.calculatePerceptualHash(`/images/${card.image?.split('/').at(-1)}`)
+                return { id: card.card_id, hash }
+              } catch (error) {
+                console.error(`Error generating hash for card ${card.card_id}:`, error)
+                return null
+              }
+            })
```

Comment:
Here if we add a new card to the promo pack, it will regenerate all the hashes of all cards. Instead you could check if the hash already exists and then skip it to save some time.